syntax = "proto3";

package message;

import "minion.proto";


message MessageData {
  bytes recipient_id = 1;
  oneof m {
    minion.HolePunchParameters hole_punch = 2;
    minion.Register minion_register = 3;
  };
};

enum SendMessageStatus {
  OK = 0;
  REJECTED = 1;
  DEST_NOT_FOUND = 2;
  FAILED = 3;
};

message SignedMessage {
  bytes sender_id = 1;
  bytes signature = 2;
  bytes sender_pubkey = 3; // optional - can be obtained
  bytes sender_cert = 4;   // optional - contains pubkey
  bytes data = 5;          // signed bytes of type MessageData
  int32 ttl = 6;
};

message SendMessageResponse {
  SendMessageStatus status = 1;
  SignedMessage message = 2;
};

service Message {
  rpc Send(SignedMessage) returns (SendMessageResponse);
  rpc SendAsync(SignedMessage) returns (SendMessageResponse);
};
